crates/matrix-sdk-crypto/src/olm/group_sessions/share_strategy.rs

```diff
@@ -0,0 +1,508 @@
+                    // set is non-empty, we must rotate.
+                    let newly_excluded =
+                        shared.difference(&recipient_device_ids).collect::<BTreeSet<_>>();
+
+                    should_rotate = !newly_excluded.is_empty();
+                    if should_rotate {
+                        // We have concluded that it should rotate, no need to continue checking
+                        // other users
+                        debug!(
+                                "Rotating a room key due to at least these devices being deleted/blacklisted {:?}",
+                                newly_excluded,
+                            );
+                        break;
+                    }
+                };
+            }
+        }
+
+        trace!(should_rotate, "Done calculating group session recipients");
+
+        Ok(CollectRecipientsResult {
+            should_rotate,
+            devices: sharing_result.allowed_devices,
+            withheld_devices: sharing_result.withheld_devices,
+        })
+    }
+
+    async fn collect_session_recipients_device_based(
+        // &self,
+        store: &Store,
+        users: BTreeSet<&UserId>,
+        only_allow_trusted_devices: bool,
+    ) -> OlmResult<KeySharingResult> {
+        let mut allowed_devices: BTreeMap<OwnedUserId, Vec<ReadOnlyDevice>> = Default::default();
+        let mut withheld_devices: Vec<(ReadOnlyDevice, WithheldCode)> = Default::default();
+
+        trace!(?users, "Calculating group session recipients - device based");
+
+        let own_identity =
+            store.get_user_identity(store.user_id()).await?.and_then(|i| i.into_own());
+
+        for user_id in users {
+            let user_devices = store.get_readonly_devices_filtered(user_id).await?;
+
+            // We only need the user identity if only_allow_trusted_devices is set.
+            let device_owner_identity = if only_allow_trusted_devices {
+                store.get_user_identity(user_id).await?
+            } else {
+                None
+            };
+
+            // From all the devices a user has, we're splitting them into two
+            // buckets, a bucket of devices that should receive the
+            // room key and a bucket of devices that should receive
+            // a withheld code.
+            let (recipients, withheld_recipients): (
+                Vec<ReadOnlyDevice>,
+                Vec<(ReadOnlyDevice, WithheldCode)>,
+            ) = user_devices.into_values().partition_map(|d| {
+                if d.is_blacklisted() {
+                    Either::Right((d, WithheldCode::Blacklisted))
+                } else if only_allow_trusted_devices
+                    && !d.is_verified(&own_identity, &device_owner_identity)
+                {
+                    Either::Right((d, WithheldCode::Unverified))
+                } else {
+                    Either::Left(d)
+                }
+            });
+
+            if !recipients.is_empty() {
```

Comment:
It was annoying in the tests to have these empty buckets of devices. Also was not usefull because we checked earlier for removed users so no need to maintain an empty bucket when looking for left devices.